app/gui2/ydoc-server/languageServerSession.ts

```diff
@@ -341,51 +347,69 @@ class ModulePersistence extends ObservableV2<{ removed: () => void }> {
+    if (result) return result
   }
 
-  async open() {
-    this.queuedAction = LsAction.Open
-    switch (this.state) {
-      case LsSyncState.Disposed:
-      case LsSyncState.WritingFile:
-      case LsSyncState.Synchronized:
-      case LsSyncState.WriteError:
-      case LsSyncState.Reloading: {
-        return
-      }
-      case LsSyncState.Closing: {
-        await this.lastAction
-        if (this.queuedAction === LsAction.Open) await this.open()
-        return
-      }
-      case LsSyncState.Opening: {
-        await this.lastAction
-        return
-      }
-      case LsSyncState.Closed: {
-        await this.withState(LsSyncState.Opening, async () => {
-          const promise = this.ls.openTextFile(this.path)
-          this.setLastAction(promise.catch(() => this.setState(LsSyncState.Closed)))
-          const result = await promise
-          if (!result.writeCapability) {
-            console.error('Could not acquire write capability for module:', this.path)
-            throw new Error(
-              `Could not acquire write capability for module '${this.path.segments.join('/')}'`,
-            )
+  async open(): Promise<Result<void>> {
+    return await withContext(
+      () => `When opening module ${this.path}`,
+      async () => {
+        this.queuedAction = LsAction.Open
+        switch (this.state) {
+          case LsSyncState.Disposed:
+          case LsSyncState.WritingFile:
+          case LsSyncState.Synchronized:
+          case LsSyncState.WriteError:
+          case LsSyncState.Reloading: {
+            return Ok()
           }
-          this.syncFileContents(result.content, result.currentVersion)
-        })
-        return
-      }
-      default: {
-        this.state satisfies never
-        return
-      }
-    }
+          case LsSyncState.Closing: {
+            await this.lastAction
+            if (this.queuedAction === LsAction.Open) return await this.open()
+            return Ok()
+          }
+          case LsSyncState.Opening: {
+            await this.lastAction
+            return Ok()
+          }
+          case LsSyncState.Closed: {
+            await this.withState(LsSyncState.Opening, async () => {
+              const promise = this.ls.openTextFile(this.path)
+              this.setLastAction(
+                promise.then((res) => !res.ok && this.setState(LsSyncState.Closed)),
+              )
+              const result = await promise
+              if (!result.ok) return result
+              if (!result.value.writeCapability) {
+                return Err(
+                  `Could not acquire write capability for module '${this.path.segments.join('/')}'`,
+                )
+              }
+              this.syncFileContents(result.value.content, result.value.currentVersion)
+              return Ok()
+            })
+            return Ok()
+          }
+          default: {
+            this.state satisfies never
+            return Ok()
```

Comment:
I know it was originally like that, but it would be better to use `assertNever` here. Now we have it in `shared`, so it shouldn't be a problem to change.